[patreon] add support for patreon-hosted video embeds

putnam opened the pull request:
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x ] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x ] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x ] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x ] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x ] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x ] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x ] Bug fix
- [x ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

The current Patreon extractor does not support videos that are embedded/encoded/hosted by Patreon itself. It only supports videos that are attached to posts, like file uploads. Patreon-native videos use a hash check on the CDN URL, requiring the user to log in for paywalled content. I have re-enabled (and updated/fixed) the login code, and added support for these types of embeds.

putnam commented:
this fixes #26940

dstftw commented on youtube_dl/extractor/patreon.py:
```diff
@@ -11,11 +11,16 @@
     parse_iso8601,
     str_or_none,
     try_get,
+    sanitized_Request,
+    ExtractorError,
```
Alphabetic.

dstftw commented on youtube_dl/extractor/patreon.py:
```diff
@@ -11,11 +11,16 @@
+    sanitized_Request,
+    ExtractorError,
 )
+import json
+import re
```
Move before youtube-dl imports.

dstftw commented on youtube_dl/extractor/patreon.py:
```diff
@@ -63,25 +68,32 @@ class PatreonIE(InfoExtractor):
 
         request = sanitized_Request(
-            'https://www.patreon.com/processLogin',
-            compat_urllib_parse_urlencode(login_form).encode('utf-8')
+            'https://www.patreon.com/api/login?include=campaign,user_location&json-api-version=1.0',
+            json.dumps(login_form).encode('ascii')
         )
```
Move this inside `_download_webpage` call.

dstftw commented on youtube_dl/extractor/patreon.py:
```diff
@@ -63,25 +68,32 @@ class PatreonIE(InfoExtractor):
         request = sanitized_Request(
-            'https://www.patreon.com/processLogin',
-            compat_urllib_parse_urlencode(login_form).encode('utf-8')
+            'https://www.patreon.com/api/login?include=campaign,user_location&json-api-version=1.0',
```
Move query to `query` parameter.

putnam commented:
thanks for the quick review. fixed the convention issue and just dropped using sanitized_Request which is no longer needed with the more robust download_webpage helper function. also there is actually no need for the query params on the patreon login page so i dropped those.

dstftw commented:
For further review you have to provide account credentials with access to paywalled video.

putnam commented:
Can you provide more specific instructions? Posting private paid Patreon
account to a github page is not a good idea.

On Wed, Jan 27, 2021 at 2:33 PM Sergey M. <notifications@github.com> wrote:

> For further review you have to provide account credentials with access to
> paywalled video.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ytdl-org/youtube-dl/pull/27917#issuecomment-768559064>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAASBT2O4ZRJO32SFCDRSYLS4B2CDANCNFSM4WOBWCJQ>
> .
>

jeLee6gi commented:
I tested your paywalled test case providing patreon cookies with `--cookies` and I get a 404 from vimeo. I am subscribed to the creator of that video and I can view it in my webbrowser.
